hooks/tk-multi-publish2/basic/nuke/publish_files.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import pprint
+import nuke
+import sgtk
+from sgtk.platform import engine
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class NukePublishFilesCustomPlugin(HookBaseClass):
+    """
+    Inherits from NukePublishFilesPlugin
+    """
+    def __init__(self, parent, **kwargs):
+        """
+        Construction
+        """
+        # call base init
+        super(NukePublishFilesCustomPlugin, self).__init__(parent, **kwargs)
+
+    def _sync_frame_range(self):
+        """
+        Checks whether frame range is in sync with shotgun.
+
+        :return: True if yes false otherwise
+        """
+        # get the current engine
+        engine = sgtk.platform.current_engine()
+        # context from engine
+        context = engine.context
+        entity = context.entity
+
+        sg_entity_type = context.entity["type"]
+        sg_filters = [["id", "is", entity["id"]]]
+        fields = ["cut_in", "cut_out"]
+
+        # get the field information from shotgun based on Shot
+        data = engine.shotgun.find_one(sg_entity_type, filters=sg_filters, fields=fields)
+        # compare if the frame range set at root level is same as the shogun cut_in, cut_out
+        root = nuke.Root()
+        if root.firstFrame() != data["cut_in"] or root.lastFrame() != data["cut_out"]:
+            self.logger.error("Frame range not synced with Shotgun.")
+            return False
+        return True
+
+    def _non_sgtk_writes(self):
+        """
+        Checks for non SGTK write nodes present in the scene.
+
+        :return: True if yes false otherwise
+        """
+        write_nodes = ""
+        # get all write nodes
+        write = nuke.allNodes('Write')
+        if write:
+            for item in range(len(write)):
+                write_nodes += "\n" + write[item]['name'].value()
+            self.logger.error("Non SGTK write nodes detected.",
+                              extra={
+                                  "action_show_more_info": {
+                                      "label": "Show Info",
+                                      "tooltip": "Show non sgtk write node(s)",
+                                      "text": "Non SGTK write nodes:\n{}".format(write_nodes)
+                                  }
+                              }
+                              )
```

Comment:
I disagree with this approach. We should just modify the Nuke collector to skip non WriteTank write nodes. Then these will never even get added to the publish process in the first place.

Reply:
A bit of context here. We were discussing with Mukesh if there will be cases (including precomp)  where non sgtk write nodes would be used instead of WriteTank nodes. Since WriteTank node handles all the requirements, reached a collective conclusion that artists can use non sgtk  write nodes during their work if they want but during publish, to keep the workflow consistent and to follow a common template we will mandate only WriteTank nodes.

We will run this through roto/paint and if they feel necessary, will relax this.

Reply:
Ok, Ashish filled me in. I think we should still modify the Nuke collector though, and that should be part of this commit.